_posts/2021-03-11-dask_memory_usage.md

```diff
@@ -0,0 +1,273 @@
+
+So after I'd left that job, I created [a memory profiler called Fil](https://pythonspeed.com/fil) that is expressly designed for finding peak memory usage.
+Unlike `dask-memusage`, which can be run on production workloads, Fil slows down your execution and has other limitations I'm currently working on (it doesn't support multiple processes, as of March 2021), so for now it's better used for manual profiling.
+
+We can write a little script that only runs on `big.txt`:
+
+```python
+from pathlib import Path
+from example2 import calculate_top_10
+
+calculate_top_10(Path("files/big.txt"))
+```
+
+Run it under Fil:
+
+```shell-session
+$ pip install filprofiler
+$ fil-profile run example3.py
+```
+
+And the result shows us where the bulk of the memory is being allocated:
+
+<iframe id="peak" src="/images/dask_memusage/peak-memory.svg" width="100%" height="300" scrolling="auto" frameborder="0"></iframe>
+
+Reading in the file takes 8% of memory, but `data.split()` is responsible for 84% of memory.
+Perhaps we shouldn't be loading the whole file into memory and splitting the whole file into words, and instead we should be processing the file line by line.
+A good next step if this were real code would be to fix the way `calculate_top_10()` is implemented.
+
+## Next steps
+
+What should you do if your Dask workload is using too much memory?
+
+If you're running Dask workloads with the Distributed backend, and you're fine with only having one thread per worker, running with `dask-memusage` will give you real-world per-task memory usage on production workloads.
+You can then use the resulting information in a variety of ways:
+
+* As a starting point for optimizing memory usage.
+  Once you know which tasks use the most memory, you can then [use Fil to figure out which lines of code are responsible](https://pythonspeed.com/articles/memory-profiler-data-scientists/) and then use [a variety of techniques to reduce memory usage](https://pythonspeed.com/articles/data-doesnt-fit-in-memory/).
+* When possible, you can fine tune your chunking size; smaller chunks will use less memory.
+  If you're using Dask Arrays you can [set the chunk size](https://docs.dask.org/en/latest/array-chunks.html); with Dask Dataframes you can [ensure good partition sizes](https://docs.dask.org/en/latest/dataframe-best-practices.html#repartition-to-reduce-overhead).
+* You can fine tune your hardware configuration, so you're not wasting RAM or CPU cores.
+  For example, on AWS you can [choose a variety of instance sizes](https://instances.vantage.sh/) with different RAM/CPU ratios, one of which may match your workload characteristics.
+
+In my original use case, the gene sequencing pipeline, I was able to use a combination of lower memory use and lower CPU use to reduce costs to a much more modest level.
```

Comment:
Love it, this feels so much more satisfying to end this way

Reply:
It's been two-three years, and I left that job two years ago, so hopefully my memory is not too inaccurate.